### PR TITLE
Trait-level flow control

### PR DESCRIPTION
**Pull request by roxanneskelly**
Checkpoint for DNM placeholder pull request.

### CODE REVIEW

**Comment by roxanneskelly** [comment]
https://highfidelity.manuscript.com/f/cases/20377

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16117/HighFidelity-Beta-PR14633-3a28204.apk

**Comment by howard-stearns** [comment]
This is one of five fixes that are being evaluated against server-under-load issues:
https://github.com/highfidelity/hifi/pull/14601
https://github.com/highfidelity/hifi/pull/14621
https://github.com/highfidelity/hifi/pull/14633
https://github.com/highfidelity/hifi/pull/14627
https://github.com/highfidelity/hifi/pull/14558
Please Do Not Merge until either:
- all five have been evaluated for how they do under server load, or
- you need the fix right away and check with Brad

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16117/HighFidelity-Beta-PR14633-3a28204.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16117/HighFidelity-Beta-Interface-PR14633-3a28204.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16129/HighFidelity-Beta-PR14633-da70271.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16143/HighFidelity-Beta-PR14633-4fc263f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16143/HighFidelity-Beta-Interface-PR14633-4fc263f.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16158/HighFidelity-Beta-PR14633-c9e6d27.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16170/HighFidelity-Beta-PR14633-2e457e2.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16170/HighFidelity-Beta-PR14633-2e457e2.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16170/HighFidelity-Beta-Interface-PR14633-2e457e2.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16183/HighFidelity-Beta-PR14633-9dfb07c.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16183/HighFidelity-Beta-PR14633-9dfb07c.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16183/HighFidelity-Beta-Interface-PR14633-9dfb07c.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16195/HighFidelity-Beta-PR14633-3d8a323.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16195/HighFidelity-Beta-Interface-PR14633-3d8a323.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16195/HighFidelity-Beta-PR14633-3d8a323.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16195/HighFidelity-Beta-Interface-PR14633-3d8a323.exe

**Comment by roxanneskelly** [comment]
Test Plan:
* On a large server (distributed3), install master.
* Run as many copies of the AC Script: https://raw.githubusercontent.com/roxanneskelly/hifi-tests/master/stress/traits/traitstest.js as can be run without the avatar mixer crashing immediately.
(i.e. run 100 copies...if it crashes, back off to 40, etc. to zero in on the max one can run without the avatar mixer failing)

* Install the PR on the server
* Run as many copies of the above script as can be run without crashing of the avatar mixer.

Expected: The PR should allow more copies to run without crashing.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16220/HighFidelity-Beta-PR14633-443b54d.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16220/HighFidelity-Beta-Interface-PR14633-443b54d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16220/HighFidelity-Beta-PR14633-443b54d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16220/HighFidelity-Beta-Interface-PR14633-443b54d.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16225/HighFidelity-Beta-PR14633-2b1547b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16225/HighFidelity-Beta-Interface-PR14633-2b1547b.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16235/HighFidelity-Beta-PR14633-2b1547b.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16235/HighFidelity-Beta-PR14633-2b1547b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16235/HighFidelity-Beta-Interface-PR14633-2b1547b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16235/HighFidelity-Beta-PR14633-2b1547b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16235/HighFidelity-Beta-Interface-PR14633-2b1547b.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16255/HighFidelity-Beta-PR14633-4f9cce1.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16255/HighFidelity-Beta-PR14633-4f9cce1.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16255/HighFidelity-Beta-Interface-PR14633-4f9cce1.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16255/HighFidelity-Beta-PR14633-4f9cce1.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16255/HighFidelity-Beta-Interface-PR14633-4f9cce1.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16276/HighFidelity-Beta-PR14633-6d791a8.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16276/HighFidelity-Beta-PR14633-6d791a8.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16276/HighFidelity-Beta-Interface-PR14633-6d791a8.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16278/HighFidelity-Beta-PR14633-0a76554.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16278/HighFidelity-Beta-Interface-PR14633-0a76554.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16278/HighFidelity-Beta-PR14633-0a76554.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16278/HighFidelity-Beta-Interface-PR14633-0a76554.exe

**Comment by antoninahifi** [comment]
Currently testing

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16288/HighFidelity-Beta-PR14633-c340d47.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16288/HighFidelity-Beta-Interface-PR14633-c340d47.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16288/HighFidelity-Beta-PR14633-c340d47.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16288/HighFidelity-Beta-Interface-PR14633-c340d47.exe

File: assignment-client/src/avatars/AvatarMixerClientData.h [hunk 1]
```diff
@@ -183,8 +192,13 @@ class AvatarMixerClientData : public NodeData {
...
     TraitsCheckTimestamp _lastReceivedTraitsChange;
 
+    AvatarTraits::TraitMessageSequence _currentTraitsMessageSequence{ 0 };
+
+    std::unordered_map<AvatarTraits::TraitMessageSequence, PerNodeTraitVersions> _perNodePendingTraitVersions;
+
     std::unordered_map<Node::LocalID, TraitsCheckTimestamp> _lastSentTraitsTimestamps;
-    std::unordered_map<Node::LocalID, AvatarTraits::TraitVersions> _sentTraitVersions;
+    PerNodeTraitVersions _perNodeSentTraitVersions;
+    PerNodeTraitVersions _perNodeAckedTraitVersions;
```
Comment (SimonWalton-HiFi) [inline assignment-client/src/avatars/AvatarMixerClientData.h]: Is it possible to add some comments about the meaning of the three new maps - I'm having difficulty following the logic. I guess _perNodePendingTraitVersions maps a trait packet sequence number and an originating node (ie avatar) to a set of trait types. The other PerNodeTraitVersions map originating nodes to traits sent for them?

File: assignment-client/src/avatars/AvatarMixerSlave.cpp [hunk 2]
```diff
@@ -80,10 +80,31 @@ int AvatarMixerSlave::sendIdentityPacket(NLPacketList& packetList, const AvatarM
     }
 }
 
+qint64 AvatarMixerSlave::addTraitsNodeHeader(AvatarMixerClientData* listeningNodeData,
+                                             const AvatarMixerClientData* sendingNodeData,
+                                             NLPacketList& traitsPacketList,
+                                             qint64 bytesWritten) {
+    if (bytesWritten == 0) {
+        if (traitsPacketList.getNumPackets() == 0) {
+            // This is the beginning of the traits packet, write out the sequence number.
```
Comment (SimonWalton-HiFi) [inline assignment-client/src/avatars/AvatarMixerSlave.cpp]: The traits packet list can have traits for several sending nodes; won't this only write the sending node ID for the first one?

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16315/HighFidelity-Beta-PR14633-d54fec4.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16315/HighFidelity-Beta-Interface-PR14633-d54fec4.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16315/HighFidelity-Beta-PR14633-d54fec4.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16315/HighFidelity-Beta-Interface-PR14633-d54fec4.exe

File: assignment-client/src/avatars/AvatarMixerSlave.cpp (hunk 2 above)
Comment (roxanneskelly) [reply assignment-client/src/avatars/AvatarMixerSlave.cpp]: The 'getNumPackets' check causes it to only write the sequence number before the first one.
The other option would be to always write the sequence number to the packet list when the packet list is created, but that might result in a packet with only a sequence number, and no traits, which would be additional network traffic.

I could move the check for an 'empty' traits list elsewhere, if that would make more sense.

The 'bytesWritten' check causes it to write the sending node only before the first trait.

File: assignment-client/src/avatars/AvatarMixerClientData.h (hunk 1 above)
Comment (roxanneskelly) [reply assignment-client/src/avatars/AvatarMixerClientData.h]: fixing

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16318/HighFidelity-Beta-PR14633-910c359.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16318/HighFidelity-Beta-PR14633-910c359.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16318/HighFidelity-Beta-Interface-PR14633-910c359.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16318/HighFidelity-Beta-PR14633-910c359.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16318/HighFidelity-Beta-Interface-PR14633-910c359.exe

**Comment by howard-stearns** [comment]
Removing the DNM that was placed until the "five fixes" were concurrency-tested.
Specifically, PREVENTS DROPS under load.

**Comment by antoninahifi** [comment]
PASS
Tested with 400 bots https://raw.githubusercontent.com/roxanneskelly/hifi-tests/master/stress/traits/traitstest.js and rWatch in distr3 - no avatar mixer crashes or disconnects were observed. Master 9814 with same bot script and on same domain is crashing on 260 bots  (250 works fine).

**Pull request merged**